### ISSUE TITLE
Updating private dicom tags using add or update method is not working as expected

### ISSUE THREAD

**Issue by svati-jain** [fo-dicom/fo-dicom#570]
### Expected behavior
When updating a private dicom tag using add or update method, it should update the existing pair with new value.

### Actual behavior
while updating private dicom tags it is adding a new entry with the same tag. 
So, after updating exiting tag value pair is there and new tag also added 
ex: exiting pair
0031,0010 - abc
after using add or update method to update 0031,0010 private tag
there are two entries in the file:-
0031,0010 - abc
0031,0010 - xyz

### Steps to reproduce the behavior
1. Load dicom file.
2. update private dicom tag using add or update method.
3. Save the dicom file
4. open output file it has two entries of same tag.

### fo-dicom version and OS/platform
3.0.2  - WIndows 7 64-bit

**Comment by anders9ustafsson** [comment]
Please provide a code snippet, in particular illustrating how you create the private tag that you are trying to update. 

I am guessing that you have not added a private creator since you are getting element numbers in the 00xx range, but I would like to be sure.

**Comment by anders9ustafsson** [comment]
@svati-jain Ignore the previous comment. The problem is that you are trying to update a private creator tag (gggg,00xx). This tag only defines the private creator associated with xx, and is automatically created when you add a private tag with a new private creator. For more info on private tags see [here](http://dicom.nema.org/medical/dicom/current/output/chtml/part05/sect_7.8.html).

To facilitate further discussion, please provide a code snipped showing what you have done so far. It will be easier to discuss around that.

**Comment by svati-jain** [comment]
Not only with private creator for any private dicom tag it is duplicating the pair

**Comment by anders9ustafsson** [comment]
Please show your code and we can discuss from there.

**Comment by svati-jain** [comment]
DicomFile dicomfile = DicomFile.Open(@"C:\windows\temp\test.dcm");
            DicomDataset dataSet = dicomfile.Dataset;
            List<DicomVR> vr = new List<DicomVR>();
            vr.Add(DicomVR.AE);

            DicomDictionary dicomDictionary = new DicomDictionary();
            dicomDictionary.Add(new DicomDictionaryEntry(new DicomTag(0002, 0017), "xyz", "keyword", DicomVM.VM_1, false, vr.ToArray()));
            dataSet.Add(DicomTag.SendingApplicationEntityTitle, "abc");
            dataSet.Add(DicomTag.ReceivingApplicationEntityTitle, "abc");


            var privateTagGroup = "0x0031";
            ushort aPrivateTag = 0;
            if (!string.IsNullOrEmpty(privateTagGroup))
            {
                aPrivateTag = Convert.ToUInt16(privateTagGroup, 16);
            }
           

            var aSnapShotTag = new DicomTag(aPrivateTag, 0x1001);
            dataSet.AddOrUpdate(new DicomLongText(aSnapShotTag, DicomEncoding.Default, "asc"));

            var aDicomVersion = "1.5.0.0";
            if (string.IsNullOrEmpty(aDicomVersion)) return;
            var aDicomVersionTag = new DicomTag(aPrivateTag, 0x1002);
            dataSet.AddOrUpdate(new DicomLongString(aDicomVersionTag, DicomEncoding.Default, aDicomVersion));
	    dicomfile.Save(@"C:\windows\temp\output.dcm");

**Comment by anders9ustafsson** [comment]
@svati-jain As far as I can see, you are not using any of the private creator `DicomTag` constructors, so the way *fo-dicom* sees it, you are **not** generating a private tag. It is not sufficient to specify an odd group number, you also need to specify a private creator string or object.

However, you have still drawn attention to an inconsistency bug in `DicomDataset`. Below is a code snippet and corresponding output to outline both intended and un-intended behavior:

    var dataset = new DicomDataset();

    // By using the .Add(DicomTag, ...) method, private tags get automatically updated so that a private
    // creator group number is generated (if private creator is new) and inserted into the tag element.
    dataset.Add<int>(new DicomTag(0x3001, 0x08, "PRIVATE"), 1);
    dataset.Add<double>(new DicomTag(0x3001, 0x12, "PRIVATE"), 3.14);
    dataset.Add<string>(new DicomTag(0x3001, 0x08, "ALSOPRIVATE"), "COOL");
    dataset.WriteToConsole();

    // As long as AddOrUpdate(DicomTag, ...) is used, the private tag is correctly identified using the
    // private creator string and the low-byte part of the element number (0x08)
    dataset.AddOrUpdate<int>(new DicomTag(0x3001, 0x08, "PRIVATE"), 100);
    dataset.WriteToConsole();

    // Confirmation that the correct item is updated, distinguished by the private creator string.
    dataset.AddOrUpdate<int>(new DicomTag(0x3001, 0x08, "ALSOPRIVATE"), -1);
    dataset.WriteToConsole();

    // ERROR: Element of the tag is not updated to include the private creator group number!
    // This will happen when using the .AddOrUpdate(DicomItem) overload.
    dataset.AddOrUpdate(new DicomIntegerString(new DicomTag(0x3001, 0x08, "PRIVATE"), 50));
    dataset.WriteToConsole();

    // WORKS! Need to explicitly call .GetPrivateTag to set tag element correctly when
    // adding/updating a DicomItem such as DicomCodeString.
    dataset.AddOrUpdate(new DicomCodeString(dataset.GetPrivateTag(new DicomTag(0x3001, 0x08, "ALSOPRIVATE")), "HOT"));
    dataset.WriteToConsole();

This will yield the following output (somewhat modified for increased clarity):

>(3001,0010) LO [PRIVATE] 
(3001,0011) LO [ALSOPRIVATE]
(3001,1108:ALSOPRIVATE) AE [COOL]
(3001,1008:PRIVATE) IS [1]
(3001,1012:PRIVATE) FD 3.14

>(3001,0010) LO [PRIVATE]
(3001,0011) LO [ALSOPRIVATE]
(3001,1108:ALSOPRIVATE) AE [COOL]
(3001,1008:PRIVATE) IS [100]
(3001,1012:PRIVATE) FD 3.14

>(3001,0010) LO [PRIVATE]
(3001,0011) LO [ALSOPRIVATE]
(3001,1108:ALSOPRIVATE) IS [-1]
(3001,1008:PRIVATE) IS [100]
(3001,1012:PRIVATE) FD 3.14

>(3001,0010) LO [PRIVATE]
(3001,0011) LO [ALSOPRIVATE]
(3001,1108:ALSOPRIVATE) IS [-1]
(3001,0008:PRIVATE) IS [50]
(3001,1012:PRIVATE) FD 3.14

>(3001,0010) LO [PRIVATE]
(3001,0011) LO [ALSOPRIVATE]
(3001,1108:ALSOPRIVATE) CS [HOT]
(3001,0008:PRIVATE) IS [50]
(3001,1012:PRIVATE) FD 3.14

(The private tags are shown out of order with the latest NuGet packages 3.0.2 and 3.1.0-alpha, but this issue has recently been fixed in PR #566.)

So in short, the `DicomDataset.Add(params DicomItem[])` and corresponding `AddOrUpdate` methods currently do not update the private tag element correctly before inserting in the dataset. For private tag update to work sufficiently, you should either:

* update the private tag explicitly using `DicomDataset.GetPrivateTag`, or
* use any of the `DicomDataset.Add<T>(DicomTag, ...)` or `DicomDataset.AddOrUpdate<T>(DicomTag, ...)` overloads, where `T` is a primitive type (`int, `string` etc.).

Unless you have registered the private tag in the dictionary, the latter alternative will not provide you with control over the applied value representation though; *fo-dicom* will apply the first fitting value representation it can find.
